crypto/keys/signer_factory.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package keys
+
+import (
+	"context"
+	"crypto"
+	"fmt"
+
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+	"github.com/google/trillian/crypto/keyspb"
+)
+
+// ProtoHandler uses the information in a protobuf message to obtain a crypto.Signer.
+// For example, the protobuf message may contain a key or identify where a key can be found.
+type ProtoHandler func(context.Context, proto.Message) (crypto.Signer, error)
+
+// ProtoGenerator creates a new private key based on a key specification.
+// It returns a proto that can be passed to a ProtoHandler to get a crypto.Signer.
+type ProtoGenerator func(context.Context, *keyspb.Specification) (proto.Message, error)
+
+// SignerFactory produces a crypto.Signer from a protobuf message describing a key.
+// If SignerFactory.Generate != nil, it can also generate new private keys.
+type SignerFactory struct {
+	// Generate creates a new private key based on a key specification.
+	// It returns a proto that can be passed to NewSigner() to get a crypto.Signer.
+	// If nil, key generation will not be possible.
+	Generate ProtoGenerator
+
+	// handlers convert a protobuf message into a crypto.Signer.
+	handlers map[string]ProtoHandler
```

Comment:
We didn't seem to shy away from global state in other places, like the [hashers registry](https://github.com/google/trillian/blob/master/merkle/hashers/tree_hasher.go#L61),  and an init-based registry would reduce the number of `+build pkcs11` files reaching for signerFactories, so it seemed worth to suggest it.

That said, your counter-arguments and the wariness about global state are certainly strong points (I found number "2" above particularly convincing). I'm happy going forward the way it is. Thanks for the detailed response, Rob.